Userland/Applications/Assistant/FuzzyMatch.cpp

```diff
@@ -0,0 +1,170 @@
+
+            // "Copy-on-Write" srcMatches into matches
+            if (first_match && srcMatches) {
+                memcpy(matches, srcMatches, nextMatch);
+                first_match = false;
+            }
+
+            // Recursive call that "skips" this match
+            uint8_t recursiveMatches[256];
+            int recursiveScore;
+            if (fuzzy_match_recursive(pattern, str + 1, recursiveScore, strBegin, matches, recursiveMatches, sizeof(recursiveMatches), nextMatch, recursionCount, recursionLimit)) {
+
+                // Pick best recursive score
+                if (!recursiveMatch || recursiveScore > bestRecursiveScore) {
+                    memcpy(bestRecursiveMatches, recursiveMatches, 256);
+                    bestRecursiveScore = recursiveScore;
+                }
+                recursiveMatch = true;
+            }
+
+            // Advance
+            matches[nextMatch++] = (uint8_t)(str - strBegin);
+            ++pattern;
+        }
+        ++str;
+    }
+
+    // Determine if full pattern was matched
+    bool matched = *pattern == '\0' ? true : false;
+
+    // Calculate score
+    if (matched) {
+        const int sequential_bonus = 15;   // bonus for adjacent matches
+        const int separator_bonus = 30;    // bonus if match occurs after a separator
+        const int camel_bonus = 30;        // bonus if match is uppercase and prev is lower
+        const int first_letter_bonus = 20; // bonus if the first letter is matched
+
+        const int leading_letter_penalty = -5;      // penalty applied for every letter in str before the first match
+        const int max_leading_letter_penalty = -15; // maximum penalty for leading letters
+        const int unmatched_letter_penalty = -1;    // penalty for every letter that doesn't matter
+
+        // Iterate str to end
+        while (*str != '\0')
+            ++str;
+
+        // Initialize score
+        outScore = 100;
+
+        // Apply leading letter penalty
+        int penalty = leading_letter_penalty * matches[0];
+        if (penalty < max_leading_letter_penalty)
+            penalty = max_leading_letter_penalty;
+        outScore += penalty;
+
+        // Apply unmatched penalty
+        int unmatched = (int)(str - strBegin) - nextMatch;
+        outScore += unmatched_letter_penalty * unmatched;
+
+        // Apply ordering bonuses
+        for (int i = 0; i < nextMatch; ++i) {
+            uint8_t currIdx = matches[i];
+
+            if (i > 0) {
+                uint8_t prevIdx = matches[i - 1];
+
+                // Sequential
+                if (currIdx == (prevIdx + 1))
+                    outScore += sequential_bonus;
+            }
+
+            // Check for bonuses based on neighbor character value
+            if (currIdx > 0) {
+                // Camel case
+                char neighbor = strBegin[currIdx - 1];
+                char curr = strBegin[currIdx];
+                if (::islower(neighbor) && ::isupper(curr))
```

Comment:
There's no other function named `islower` or `isupper` in this scope, I'd say the `::` is redundant.